src/audio/dcblock/dcblock_hifi4.c

```diff
@@ -0,0 +1,173 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright(c) 2022 Intel Corporation. All rights reserved.
+//
+// Author: Andrula Song <andrula.song@intel.com>
+
+#include <stdint.h>
+#include <sof/audio/component.h>
+#include <sof/audio/format.h>
+#include <sof/audio/dcblock/dcblock.h>
+
+#if defined(__XCC__) && XCHAL_HAVE_HIFI4
```

Comment:
Keep all the logic and all decisions all visible and centralized in a single place; in the common .h file. Keep the .c files as "cpp-dumb" as possible with a single `#if CONSTANT`.

Reply:
this also needs to include the compiler_info header

Reply:
> `defined(__XCC__)`

Is this defined for the newer `xt-clang` too or only for the old `xt-xcc`?

Reply:
@andyross do you know?

Reply:
With help from @dcpleung 

```shell
find XCC/ -name xt-xcc -o -name xt-clang | while read -r XT
   do printf "\n\n --- $XT --- \n\n" 
   </dev/null $XT -E -dM - | grep -i -e xcc -e clang
done

 --- XCC/install/tools/RD-2012.5-linux/XtensaTools/bin/xt-xcc --- 

#define __XCC__ 9000
#define __XCC_MINOR__ 5


 --- XCC/install/tools/RF-2016.4-linux/XtensaTools/bin/xt-xcc --- 

#define __XCC__ 11000
#define __XCC_MINOR__ 4


 --- XCC/install/tools/RG-2017.8-linux/XtensaTools/bin/xt-xcc --- 

#define __XCC__ 12000
#define __XCC_MINOR__ 8

 --- XCC/install/tools/RG-2017.8-linux/XtensaTools/bin/xt-clang --- 

XCC/install/tools/RG-2017.8-linux/XtensaTools/bin/xt-clang: error while loading shared libraries:
libncurses.so.5: cannot open shared object file: No such file or directory


 --- XCC/install/tools/RI-2020.5-linux/XtensaTools/bin/xt-xcc --- 

#define __XCC__ 14000
#define __XCC_MINOR__ 5

 --- XCC/install/tools/RI-2020.5-linux/XtensaTools/bin/xt-clang --- 

XCC/install/tools/RI-2020.5-linux/XtensaTools/llvm/bin/clang: error while loading shared libraries:
libtinfo.so.5: cannot open shared object file: No such file or directory


 --- XCC/install/tools/RI-2022.10-linux/XtensaTools/bin/xt-clang --- 

#define __CLANG_ATOMIC_BOOL_LOCK_FREE 2
#define __CLANG_ATOMIC_CHAR16_T_LOCK_FREE 2
#define __CLANG_ATOMIC_CHAR32_T_LOCK_FREE 2
#define __CLANG_ATOMIC_CHAR_LOCK_FREE 2
#define __CLANG_ATOMIC_INT_LOCK_FREE 2
#define __CLANG_ATOMIC_LLONG_LOCK_FREE 1
#define __CLANG_ATOMIC_LONG_LOCK_FREE 2
#define __CLANG_ATOMIC_POINTER_LOCK_FREE 2
#define __CLANG_ATOMIC_SHORT_LOCK_FREE 2
#define __CLANG_ATOMIC_WCHAR_T_LOCK_FREE 2
#define __VERSION__ "XtensaTools-14.10 Clang 10.0.1 "
#define __XCC_MINOR__ 10
#define __XCC__ 14000
#define __XT_CLANG_MINOR__ 10
#define __XT_CLANG__ 14000
#define __clang__ 1
#define __clang_major__ 10
#define __clang_minor__ 0
#define __clang_patchlevel__ 1
#define __clang_version__ "10.0.1 "



```

Reply:
> error while loading shared libraries

A different Linux distribution fixed that:

```
 --- xcc/install/tools/RG-2017.8-linux/XtensaTools/bin/xt-clang --- 

Error: there is no Xtensa core registered as the default.

You need to either specify the name of a registered Xtensa core (with
the --xtensa-core option or the XTENSA_CORE environment variable) or
specify a different registry of Xtensa cores (with the --xtensa-system
option or the XTENSA_SYSTEM environment variable).

 --- xcc/install/tools/RI-2020.5-linux/XtensaTools/bin/xt-clang --- 

#define __CLANG_ATOMIC_BOOL_LOCK_FREE 2
#define __CLANG_ATOMIC_CHAR16_T_LOCK_FREE 2
#define __CLANG_ATOMIC_CHAR32_T_LOCK_FREE 2
#define __CLANG_ATOMIC_CHAR_LOCK_FREE 2
#define __CLANG_ATOMIC_INT_LOCK_FREE 2
#define __CLANG_ATOMIC_LLONG_LOCK_FREE 1
#define __CLANG_ATOMIC_LONG_LOCK_FREE 2
#define __CLANG_ATOMIC_POINTER_LOCK_FREE 2
#define __CLANG_ATOMIC_SHORT_LOCK_FREE 2
#define __CLANG_ATOMIC_WCHAR_T_LOCK_FREE 2
#define __VERSION__ "4.2.1 Compatible XtensaTools-14.05 Clang 6.0.1 "
#define __XCC_MINOR__ 5
#define __XCC__ 14000
#define __XT_CLANG_MINOR__ 5
#define __XT_CLANG__ 14000
#define __clang__ 1
#define __clang_major__ 6
#define __clang_minor__ 0
#define __clang_patchlevel__ 1
#define __clang_version__ "6.0.1 "


```

Reply:
Sorry, you guys beat me to it, but for future reference all gcc variants support some variant of "-dM -E" on the command line to emit the full list of built-in preprocessor symbols, e.g.:

```
echo | xt-clang -dM -E -o - - | fgrep __XCC__
#define __XCC__ 14000
```

xt-xcc is cranky and doesn't understand "-" for standard I/O:

```
xt-xcc -dM -E -o foo.syms foo.c | fgrep __XCC__ foo.syms
#define __XCC__ 14000
```

(Both of these were with a RI-2021.6 toolchain and XTENSA_CORE=ace10_LX7HiFi4_RI_2020_5)

Reply:
And the libtinfo thing is because it's linked against ancient ncurses.  FWIW on arch I've had success just linking libtinfo.so.6 to libtinfo.so.5, nothing we do actually tries anything interactive on the terminal so I doubt we'd hit any code paths that might expose the difference.

Reply:
> > `defined(__XCC__)`
> 
> Is this defined for the newer `xt-clang` too or only for the old `xt-xcc`?

Since I did not find any xt_clang macro but XCC for our latest configuration, and checked with Chao, he suggested use XCC as usual.